code/game/machinery/dance_machine.dm

```diff
@@ -328,8 +328,9 @@
 
 /obj/machinery/disco/proc/dance(mob/living/M) //Show your moves
 	set waitfor = FALSE
-	if(M.client && !(M.client.prefs.sound & SOUND_DISCO)) //We have a client that doesn't want to dance.
+	if(M.client?.prefs.sound & PREFTOGGLE_2_DANCE_DISCO) //We have a client that doesn't want to dance but listen to audio.
 		rangers -= M //Doing that here as it'll be checked less often than in processing.
+	if(M.client && !(M.client.prefs.sound & SOUND_DISCO)) //We have a client that doesn't want to dance.
 		return
```

Comment:
Keep the same if check, just do `if((no sound pref) || (no dance pref))`

Reply:
Working with @SabreML on this one, ill quote him on the issue of doing it like this as it causes audio errors,

> Because the check in dance() is removing the user from the rangers list if they have the preference disabled, when it gets around to process() it's finding the user, checking if they're in the rangers list (which they aren't anymore), and playing the music to them again